src/ucp/core/ucp_worker.c

```diff
@@ -2325,10 +2328,18 @@ static void ucp_worker_discarded_uct_eps_cleanup(ucp_worker_h worker)
 static void ucp_worker_destroy_eps(ucp_worker_h worker)
 {
     ucp_ep_ext_gen_t *ep_ext, *tmp;
+    ucp_ep_h ep;
 
     ucs_debug("worker %p: destroy all endpoints", worker);
     ucs_list_for_each_safe(ep_ext, tmp, &worker->all_eps, ep_list) {
-        ucp_ep_disconnected(ucp_ep_from_ext_gen(ep_ext), 1);
+        ep = ucp_ep_from_ext_gen(ep_ext);
+        if (ucp_ep_config(ep)->key.err_mode == UCP_ERR_HANDLING_MODE_PEER) {
```

Comment:
why need to check err mode?

Reply:
we should do discarding (i.e. flush(CANCEL)) for UCTs which support error handling

Reply:
why not discard anyway?

Reply:
> why not discard anyway?

UCT requires to do flush(CANCEL) only for transports which support error handling
otherwise - transports (e.g. XPMEM) will fail (e.g. XPMEM fails with BUS error, because remote memory doesn't exist anymore)

Reply:
so what would happen when we don't discard lanes when err handling is off?
also, how does it work, since we already in worker destroy flow? and we won't call worker progress any more?

Reply:
> so what would happen when we don't discard lanes when err handling is off?

we don't care about it. this situation shouldn't happen

> also, how does it work, since we already in worker destroy flow? and we won't call worker progress any more?

what we need it just to make a call for flush(CANCEL), we rely on the fact that it will be completed on UD in a single call w/o INPROGRESS.

Reply:
sounds like an assumption we better not rely on. what happens on other transports?
if we just need to call flush cancel once before ep_destory (why?) - probably no need to go through all discard process?